### PR TITLE
test/BIP324: disconnection scenarios during v2 handshake

### PR DESCRIPTION
**Pull request by stratospher**
Add tests for the following v2 handshake scenarios:
1. Disconnection happens when > `MAX_GARBAGE_LEN` bytes garbage is sent
2. Disconnection happens when incorrect garbage terminator is sent
3. Disconnection happens when garbage bytes are tampered with
4. Disconnection happens when AAD of first encrypted packet after the garbage terminator is not filled
5. bitcoind ignores non-empty version packet and no disconnection happens

All these tests require a modified v2 P2P class (different from `EncryptedP2PState` used in `v2_p2p.py`) to implement our custom handshake behaviour based on different scenarios and have been kept in a single test file (`test/functional/p2p_v2_misbehaving.py`). Shifted the test in `test/functional/p2p_v2_earlykeyresponse.py` which is of the same pattern to this file too.

### CODE REVIEW

**Comment by stratospher** [comment]
@sr-gi, i tried [this manually sending idea](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) but i still think intermittent failures are possible there.
- we can't get rid of `can_data_be_received` variable because if we don't use this variable, test would succeed irrespective of whether we send 4 bytes network magic first or 4 bytes from ellswift bytes first and we don't want that.
- so since `data_received()` always happens in `Network thread`  and send of ellswift bytes + setting  `can_data_be_received=True` happens on `MainThread`, in the rare scenario that `data_received()` gets called before setting `can_data_be_received`, an intermittent failure could happen i think.

here's a [branch](https://github.com/stratospher/bitcoin/commit/7e7173467fd5459e4b19ebd48186690f174be95f) where i tweaked the code you shared a bit with a sleep statement for making the test crash and reintroducing `can_data_be_received` variable.

File: test/functional/p2p_v2_misbehaving.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        when TestType = (EXCESS_GARBAGE, WRONG_GARBAGE_TERMINATOR, WRONG_GARBAGE)"""
+        self.privkey_ours, self.ellswift_ours = ellswift_create()
+
+        if self.test_type == TestType.EXCESS_GARBAGE:
+            # send > 4095 bytes garbage
+            garbage_len = MAX_GARBAGE_LEN + random.randrange(1, 10)
+        elif self.test_type == TestType.WRONG_GARBAGE_TERMINATOR:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN//2)
+        else:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN + 1)
```
Comment (kevkevinpal) [inline test/functional/p2p_v2_misbehaving.py:60]: doesn't the `+1` make the `garbage_len` greater than `4095` which is what is being covered in `TestType.EXCESS_GARBAGE`

Comment (kevkevinpal) [reply test/functional/p2p_v2_misbehaving.py:60]: never mind randrange doesn't include the value passed

Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py:60]: yes!

**Review by kevkevinpal — COMMENTED** [review]
Concept ACK

added two comments I will continue to test these changes as I am still trying to understand fully what the tests do exactly

File: test/functional/p2p_v2_misbehaving.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            [],  # EARLY_KEY_RESPONSE
+            ["V2 transport error: missing garbage terminator, peer=1"],  # EXCESS_GARBAGE
+            ["version handshake timeout peer=2"],  # WRONG_GARBAGE_TERMINATOR
+            ["V2 transport error: packet decryption failure"],  # WRONG_GARBAGE
+            ["V2 transport error: packet decryption failure"],  # SEND_NO_AAD
+            [],  # SEND_NON_EMPTY_VERSION_PACKET
+        ]
+        for test_type in TestType:
+            if test_type == TestType.EARLY_KEY_RESPONSE:
+                continue
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py:190]: `test_earlykeyresponse` sends handshake bytes etc.. in the `MainThread` and I feel it's better to isolate that behaviour in a different test.

also, if you grep for `def test_` in `test/functional`, you'll see many [examples](https://github.com/bitcoin/bitcoin/blob/3cbc8cbc71d3d6ecfaf41164ce59c24ac94bae99/test/functional/feature_asmap.py#L128) of different tests listed in `run_test()` using the pattern which is currently used.

File: test/functional/p2p_v2_misbehaving.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+
+class EncryptedP2PMisbehaving(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [["-v2transport=1", "-peertimeout=3"]]
+
+    def run_test(self):
+        self.test_earlykeyresponse()
+        self.test_v2disconnection()
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py:158]: yeah! `expected_debug_message` on [L180](https://github.com/bitcoin/bitcoin/blob/ec9005ca4be088dfa2d247bbfe964a9c98e4f29d/test/functional/p2p_v2_misbehaving.py#L180) was written with the assumption that `test_v2disconnection` is run after `test_earlykeyresponse`. so peer=0 happens in `test_earlykeyresponse` and peer=1,2,3,4,5 happens in  `test_v2disconnection`.

**Comment by sr-gi** [comment]
> * we can't get rid of `can_data_be_received` variable because if we don't use this variable, test would succeed irrespective of whether we send 4 bytes network magic first or 4 bytes from ellswift bytes first and we don't want that.

I've been playing around sending anything different than the first 4 bytes as the network magic and test fail on my end (which is expected given `data_received` is called with `magic_sent` false on the first-byte mismatch.

Do you have any working example I can try to reproduce?

**Comment by stratospher** [comment]
> I've been playing around sending anything different than the first 4 bytes as the network magic and test fail on my end (which is expected given data_received is called with magic_sent false on the first-byte mismatch).

yes! that's expected behaviour and sending anything other than the first 4 bytes as network magic will make the test fail on master.

what i was trying to say was:
1. on master
- replacing `return b"\xfa\xbf\xb5\xda"` with any other 4 bytes will make the test fail and this is expected behaviour
3. using [this patch](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) mentioned in #29352
- replacing `return b"\xfa\xbf\xb5\xda"` with any other 4 bytes will make the test pass and this isn't behaviour we want
- this happens because [`can_data_be_received` variable](https://github.com/bitcoin/bitcoin/blob/ab5dfdbec1143f673f4d83acd4e335bb2c51034e/test/functional/p2p_v2_earlykeyresponse.py#L29) is removed in the patch

so i was just trying to say that we can't remove [`can_data_be_received` variable](https://github.com/bitcoin/bitcoin/blob/ab5dfdbec1143f673f4d83acd4e335bb2c51034e/test/functional/p2p_v2_earlykeyresponse.py#L29).

**Review by mzumsande — COMMENTED** [review]
Concept ACK

The one thing I'm a bit unsure about is that this duplicates quite a lot of code, especially in `complete_handshake` - the main implementation in `v2_p2p.py` and the overwritten version from this test could run out of sync in the future. But I don't have a good idea how to avoid this yet...

File: test/functional/p2p_v2_misbehaving.py [hunk 4]
```diff
@@ -3,87 +3,107 @@
...
-        peer1 = node0.add_p2p_connection(PeerEarlyKey(), wait_for_verack=False, send_version=False, supports_v2_p2p=True, wait_for_v2_handshake=False)
+        peer1 = node0.add_p2p_connection(MisbehavingV2Peer(TestType.EARLY_KEY_RESPONSE), wait_for_verack=False, send_version=False, supports_v2_p2p=True, wait_for_v2_handshake=False)
         self.wait_until(lambda: peer1.connection_opened)
         self.log.info('Sending remaining ellswift and garbage which are different from V1_PREFIX. Since a response is')
         self.log.info('expected now, our custom data_received() function wouldn\'t result in assertion failure')
         ellswift_and_garbage_data = peer1.v2_state.initiate_v2_handshake()
         peer1.send_raw_message(ellswift_and_garbage_data)
-        peer1.wait_for_disconnect(timeout=5)
-        self.log.info('successful disconnection when MITM happens in the key exchange phase')
+        with self.nodes[0].assert_debug_log(['version handshake timeout peer=0']):
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: done.

File: test/functional/p2p_v2_misbehaving.py [hunk 5]
```diff
@@ -3,87 +3,107 @@
...
-    - `can_data_be_received` is a variable used to assert if data is received on recvbuf.
-            - v2 TestNode shouldn't respond back if we send V1_PREFIX and data shouldn't be received on recvbuf.
-              This state is represented using `can_data_be_received` = False.
-            - v2 TestNode responds back when mismatch from V1_PREFIX happens and data can be received on recvbuf.
-              This state is represented using `can_data_be_received` = True.
+class TestType(Enum):
+    """ Scenarios to be tested:
+
+    1. EARLY_KEY_RESPONSE - The responder needs to wait until one byte is received which does not match the 16 bytes
+    consisting of network magic followed by "version\x00\x00\x00\x00\x00" before sending out it's ellswift + garbage bytes
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: done.

File: test/functional/test_framework/test_node.py [hunk 6]
```diff
@@ -694,6 +694,9 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, send_version=Tru
         supports_v2_p2p = self.use_v2transport and supports_v2_p2p
         p2p_conn.peer_connect(**kwargs, send_version=send_version, net=self.chain, timeout_factor=self.timeout_factor, supports_v2_p2p=supports_v2_p2p)()
 
+        if wait_for_disconnect:
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/test_framework/test_node.py]: woah more racy thread logic. good catch! i've implemented the `expect_success` parameter suggestion.

**Comment by stratospher** [comment]
> The one thing I'm a bit unsure about is that this duplicates quite a lot of code, especially in complete_handshake - the main implementation in v2_p2p.py and the overwritten version from this test could run out of sync in the future. But I don't have a good idea how to avoid this yet...

hmm good point. will need to think about it more. simplest solution would be keeping it all in main implementation in `v2_p2p.py` haha but doesn't make sense to have it there just for this test file :(

**Comment by sr-gi** [comment]
I find the approach fairly hard to follow here. Having all the logic in the constructor with if/elses based on the connection type instead of having different constructors/a test for each type of fail feels really error-prone and difficult to make sure if a test is doing what we expect. I think it'd be better to have multiple classes that build from `v2_p2p` and modify what's needed, even if the file gets larger.

Just an example:

In the `WRONG_GARBAGE_TERMINATOR` case, we are modifying the gargabe in two ways. First, we are making it smaller than expected, but also, we are randomly flipping some of its bits. Turns out if you removed both changes, the tests still fails, but I'm not sure why. I would expect this not to be the case, since removing the relevant parts should make the class equal to the expected `v2_p2p`.

**Comment by stratospher** [comment]
> I think it'd be better to have multiple classes that build from v2_p2p and modify what's needed, even if the file gets larger.

that's a great suggestion @sr-gi! it's possible to design those classes in such a way that it avoids code duplication and the file is almost the same size!

> Turns out if you removed both changes, the tests still passes, but I'm not sure why. 

because we don't send a version message to ensure that the disconnection happens in the v2 handshake phase. in your case (different from what we are testing here), we send the correct garbage terminator but because we don't send a version message - disconnection is still expected.

**Comment by stratospher** [comment]
I've updated the PR to have different child class implementations for `EncryptedP2PState` based on the disconnection scenario we're testing - `EarlyKeyResponseState`, `ExcessGarbageState` etc...

Also introduced 2 more commits for cleaner code:
- c642b08c - logging when the garbage is sent instead of when garbage is generated (suggestion from https://github.com/bitcoin/bitcoin/pull/24748#discussion_r1466826427 which is useful now that we have multiple child classes for `EncryptedP2PState`)
- 38eb429 - transport version as a class variable instead of a global variable

File: test/functional/p2p_v2_misbehaving.py [hunk 7]
```diff
@@ -3,87 +3,107 @@
...
+        """Send ellswift and garbage bytes in 2 parts when TestType = (EARLY_KEY_RESPONSE)"""
+        if self.test_type == TestType.EARLY_KEY_RESPONSE:
+            # Here, the 64 bytes ellswift is assumed to have it's first 4 bytes match network magic bytes.
+            # It is sent in 2 phases:
+            # 1. when `magic_sent` = False, send first 4 bytes of ellswift (matches network magic bytes)
+            # 2. when `magic_sent` = True, send remaining 60 bytes of ellswift
+            if not self.magic_sent:
+                self.generate_keypair_and_garbage()
+                self.magic_sent = True
+                return b"\xfa\xbf\xb5\xda"
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: done.

**Review by sr-gi — COMMENTED** [review]
I like this approach much better, it is way easier to follow and reason about. I left some comments inline.

> because we don't send a version message to ensure that the disconnection happens in the v2 handshake phase. in your case (different from what we are testing here), we send the correct garbage terminator but because we don't send a version message - disconnection is still expected.

What is weird is that the test is expecting a specific log message, which seems to also match in this case. This is still the case by the way. Is there an alternative way of making sure this does not happen?

File: test/functional/test_framework/v2_p2p.py [hunk 8]
```diff
@@ -79,6 +78,7 @@ class EncryptedP2PState:
 
         encrypt/decrypt v2 P2P messages using v2_enc_packet() and v2_receive_packet().
     """
+    transport_version = b''
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/test_framework/v2_p2p.py]: done.

File: test/functional/p2p_v2_misbehaving.py [hunk 9]
```diff
@@ -3,87 +3,100 @@
...
+        if not self.magic_sent:
+            self.generate_keypair_and_garbage()
+            self.magic_sent = True
+            return MAGIC_BYTES[self.net]
         else:
+            # `can_data_be_received` is a variable used to assert if data is received on recvbuf.
+            # 1. v2 TestNode shouldn't respond back if we send V1_PREFIX and data shouldn't be received on recvbuf.
+            # This state is represented using `can_data_be_received` = False.
+            # 2. v2 TestNode responds back when mismatch from V1_PREFIX happens and data can be received on recvbuf.
+            # This state is represented using `can_data_be_received` = True.
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: > I still think this can be simplified. initiate_v2_handshake is going to be called, asynchronously, when creating this object. You can reduce a lot of the boilerplate, and get rid of magic_sent, as long as you do the last bit manually on the test (or just create another method, like continue_v2_handshake).

yes, that can be done since the last bit is being sent from `MainThread` anyways.

before: 
1. send both 4 bytes network magic (first 4 bytes of ellswift) - `NetworkThread`
2. remaining `ellswift_garbage` bytes - `MainThread`

now:
1. send both 4 bytes network magic (first 4 bytes of ellswift) - `MainThread`
2. remaining `ellswift_garbage` bytes - `MainThread`

realised [my earlier comment](https://github.com/bitcoin/bitcoin/pull/29431#issuecomment-1945395946) is the situation in master too and a sleep in `MainThread` where there's no waiting for some other tasks is unlikely. 😬

I've used [your original suggestion](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) and updated the PR to send everything manually!

**Comment by stratospher** [comment]
thank you for the helpful reviews @sr-gi!

I've updated the PR:
- `EARLY_KEY_RESPONSE` test doesn't have `magic_sent` now and the `initial_v2_handshake()` function is cleaner
- `EARLY_KEY_RESPONSE` test does sending of 4 bytes ellswift (which match network magic) + remaining ellswift and garbage bytes on `MainThread` - both are done manually now
- Added v2 handshake timeout log in 7d07daa

     
> What is weird is that the test is expecting a specific log message, which seems to also match in this case. This is still the case by the way. Is there an alternative way of making sure this does not happen?

hmm fair point, i've added a log message in `InactivityCheck()` to distinguish v2 handshake timeout case from version handshake timeout.

**Review by sr-gi — COMMENTED** [review]
I love it now 🚀 

ACK ae528cb9e29b75d36a8018954f45bcfb886a4557

File: test/functional/test_framework/test_node.py [hunk 10]
```diff
@@ -687,14 +687,15 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, send_version=Tru
         if supports_v2_p2p is None:
             supports_v2_p2p = self.use_v2transport
 
-
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/test_framework/test_node.py:690]: `PEP 8: E303 too many blank lines (2)`  was showing up in my code editor. so maybe we can keep it since we're touching the function here.

File: test/functional/p2p_v2_misbehaving.py [hunk 11]
```diff
@@ -3,87 +3,79 @@
...
+        peer1 = node0.add_p2p_connection(MisbehavingV2Peer(TestType.EARLY_KEY_RESPONSE), wait_for_verack=False, send_version=False, supports_v2_p2p=True, wait_for_v2_handshake=False)
+        peer1.send_raw_message(MAGIC_BYTES['regtest'])
         self.log.info('Sending remaining ellswift and garbage which are different from V1_PREFIX. Since a response is')
         self.log.info('expected now, our custom data_received() function wouldn\'t result in assertion failure')
-        ellswift_and_garbage_data = peer1.v2_state.initiate_v2_handshake()
-        peer1.send_raw_message(ellswift_and_garbage_data)
-        peer1.wait_for_disconnect(timeout=5)
-        self.log.info('successful disconnection when MITM happens in the key exchange phase')
+        peer1.send_raw_message(peer1.v2_state.ellswift_ours[4:] + peer1.v2_state.sent_garbage)
+        peer1.v2_state.can_data_be_received = True
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py:146]: agree! done.

File: test/functional/p2p_v2_misbehaving.py [hunk 12]
```diff
@@ -32,6 +38,15 @@ def initiate_v2_handshake(self):
         return b""
 
 
+class ExcessGarbageState(EncryptedP2PState):
+    """Generate > MAX_GARBAGE_LEN garbage bytes"""
+    def generate_keypair_and_garbage(self):
+        ellswift_garbage_bytes = super().generate_keypair_and_garbage()
+        even_more_garbage = random.randbytes(MAX_GARBAGE_LEN)
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py:54]: oh looks like 0 garbage bytes being returned can also mess up the functions where we're tampering the garbage bytes (`WRONG_GARBAGE`) and `SEND_NO_AAD` where AAD is the garbage bytes sent. so I've change `generate_keypair_and_garbage()` in `EncryptedP2PState` to return 1 bytes garbage minimum instead.

File: test/functional/p2p_v2_misbehaving.py [hunk 13]
```diff
@@ -3,87 +3,79 @@
...
     def run_test(self):
+        self.test_earlykeyresponse()
+
+    def test_earlykeyresponse(self):
         self.log.info('Sending ellswift bytes in parts to ensure that response from responder is received only when')
         self.log.info('ellswift bytes have a mismatch from the 16 bytes(network magic followed by "version\\x00\\x00\\x00\\x00\\x00")')
         node0 = self.nodes[0]
         self.log.info('Sending first 4 bytes of ellswift which match network magic')
         self.log.info('If a response is received, assertion failure would happen in our custom data_received() function')
         # send happens in `initiate_v2_handshake()` in `connection_made()`
```
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: done.

File: test/functional/test_framework/test_node.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mazzika1) [reply test/functional/test_framework/test_node.py:690]: Okay

File: test/functional/p2p_v2_misbehaving.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Mazzika1) [reply test/functional/p2p_v2_misbehaving.py:54]: Okay

[the remark this answers is not in the document]
Comment (Mazzika1) [reply test/functional/p2p_v2_misbehaving.py:54]: Okay

[the remark this answers is not in the document]
Comment (mzumsande) [reply test/functional/p2p_v2_misbehaving.py:54]: Removing the possibility of 0 garbage in `EncryptedP2PState` just to make this test nicer is a bit unfortunate in my opinion: If the node had a bug dealing with zero-length garbage, the test framework wouldn't be able to catch that anymore. Can't we change it locally just for the test cases who need it?

[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_v2_misbehaving.py]: hmm good point. I've added `garbage_len` as an optional argument in `generate_keypair_and_garbage()` so that some of the test cases which don't support 0 length garbage can avoid it.

**Review by mzumsande — COMMENTED** [review]
ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

I'd prefer combining some of the first commits with the respective test commits that actually make use of the change, but that's just my opinion ad maybe a matter of taste.

**Review by theStack — APPROVED** [review]
Code-review ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

Nice approach using a `EncryptedP2PState` subclass for each test scenario, I agree with other reviewers that this looks much cleaner now and is easier to follow.

**Comment by achow101** [comment]
ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

**Comment by fanquake** [comment]
Looks like this is causing intermittent CI failures. See #30419.

**Comment by hebasto** [comment]
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.
